core/src/main/scala/org/apache/spark/ui/PagedTable.scala

```diff
@@ -0,0 +1,222 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ui
+
+import scala.xml.{Node, Unparsed}
+
+/**
+ * A data source that provides data for a page.
+ *
+ * @param page the page number
+ * @param pageSize the number of rows in a page
+ */
+private[ui] abstract class PagedDataSource[T](page: Int, pageSize: Int) {
+
+  protected val data: Seq[T]
+
+  /**
+   * Slice the data for this page
+   */
+  def pageData: PageData[T] = {
```

Comment:
No. It uses `data`. So it cannot be `val` because `data` is null in this class's constructor.